autonomous/autonomous.py

```diff
@@ -1,71 +1,114 @@
-#         for idx, rotation in self.cubes[::-1]:
-#             position = all_pieces[idx]
-#             # flip angle, position is already correctly flipped
-#             angle = field_flip_rotation2d(rotation) if is_red() else rotation
-#             # approach angle is the same as chassis heading
-#             pose = Pose2d(position, angle)
-#             self.scoring.cube_stack.append((pose, angle))
-
-#         self.scoring.score_stack = self.nodes[::-1]
-
-#     def on_iteration(self, tm: float) -> None:
-#         if not self.finished:
-#             self.scoring.autodrive = True
-#             self.scoring.engage()
-#         if len(self.scoring.cube_stack) == 0 and len(self.scoring.score_stack) == 0:
-#             self.finished = True
-
-#     def on_disable(self) -> None:
-#         self.scoring.wants_piece = GamePiece.CONE
-
-
-# class WallSide3(CubeAutoBase):
-#     MODE_NAME = "Wall side 3"
-
-#     def __init__(self):
-#         super().__init__(
-#             cubes=[(0, Rotation2d(0)), (1, Rotation2d.fromDegrees(40))],
-#             nodes=[
-#                 Node(row=Rows.HIGH, col=0),
-#                 Node(row=Rows.HIGH, col=1),
-#                 Node(row=Rows.MID, col=1),
-#             ],
-#         )
-
-
-# class SubstationSide3(CubeAutoBase):
-#     MODE_NAME = "Substation side 3"
-#     DEFAULT = True
-
-#     def __init__(self):
-#         super().__init__(
-#             cubes=[(3, Rotation2d(0)), (2, Rotation2d.fromDegrees(-40))],
-#             nodes=[
-#                 Node(row=Rows.HIGH, col=8),
-#                 Node(row=Rows.HIGH, col=7),
-#                 Node(row=Rows.MID, col=7),
-#             ],
-#         )
+from magicbot.state_machine import AutonomousStateMachine, state
+from dataclasses import dataclass
+
+from controllers.arm import Setpoint, Setpoints
+from controllers.movement import Movement
+from controllers.score_game_piece import ScoreGamePieceController
+from controllers.acquire_cube import AcquireCubeController
+
+from utilities.functions import grid_col_to_field_y
+
+from wpimath.geometry import Pose2d, Rotation2d, Translation2d
+
+
+from math import radians
+
+
+@dataclass
+class PickupPath:
+    goal: Pose2d
+    approach_direction: Rotation2d
+    intermediate_waypoints: list[Translation2d]
+
+
+@dataclass
+class ScorePath:
+    goal: Pose2d
+    approach_direction: Rotation2d
+    intermediate_waypoints: list[Translation2d]
+    arm_setpoint: Setpoint
+
+
+class AutoBase(AutonomousStateMachine):
+    movement: Movement
+    score_game_piece: ScoreGamePieceController
+    acquire_cube: AcquireCubeController
+
+    INTAKE_PRE_TIME = 2.0
+    MANUAL_CUBE_TIME = 1.5
+
+    def __init__(self) -> None:
+        self.pickup_paths: list[PickupPath] = []
+        self.score_paths: list[ScorePath] = []
+        self.progress_idx = 0
+
+    @state(first=True)
+    def score(self, initial_call: bool) -> None:
+        if initial_call:
+            self.score_game_piece.engage()
```

Comment:
You need to tell the scoring controller which node to score in, I think ScorePath should have the goal as a node rather than a Pose2d

Reply:
you will also need a way to give the scoring controller the intermediate waypoints

Reply:
Perhaps there should be an extra state just for driving, rather than making the scoring controller responsible for navigation. Much easier to understand the steps involved if it is done that way.

Reply:
Good suggestion, so the autonomous would run the movement and then start the scoring controller from the deploying_arm state?

Reply:
Yes, something like that. I think it is nicer if this controller has clear states to make it easier to understand. e.g.:
* score cone
* drive to cube
* pick up cube
* drive to grid
* score cube
* etc